src/main/java/it/eng/idsa/dataapp/web/rest/DataControllerHttpHeader.java

```diff
@@ -55,112 +49,29 @@ public ResponseEntity<?> routerHttpHeader(@RequestHeader HttpHeaders httpHeaders
 			logger.info("Payload is empty");
 		}
 
-		HttpHeaders responseHeaders = createResponseMessageHeaders(httpHeaders, null);
-		String responsePayload = null;
-		
-		if (!("ids:RejectionMessage".equals(responseHeaders.get("IDS-Messagetype").get(0)))) {
-			responsePayload = messageUtil.createResponsePayload(httpHeaders, payload);
-		}
-		if(responsePayload == null && 
-				"ids:ContractRequestMessage".equals(responseHeaders.get("IDS-Messagetype").get(0))) {
-			logger.info("Creating rejection message since contract agreement was not found");
-			responseHeaders = createResponseMessageHeaders(httpHeaders, RejectionReason.NOT_FOUND.name());
-		}	
-		
-		if(responsePayload == null && 
-				"ids:DescriptionRequestMessage".equals(responseHeaders.get("IDS-Messagetype").get(0))) {
-			logger.info("Creating rejection message since self description could not be fetched");
-			responseHeaders = createResponseMessageHeaders(httpHeaders, RejectionReason.INTERNAL_RECIPIENT_ERROR.name());
-		}	
-		
-		if (responsePayload != null && responsePayload.contains("IDS-RejectionReason")) {
-			responseHeaders = createResponseMessageHeaders(httpHeaders, responsePayload.substring(responsePayload.indexOf(":")+1));
-			responsePayload = null;
-		}
-		
-		
-		ResponseEntity<?> response = ResponseEntity.noContent()
-				.headers(responseHeaders)
-				.build();
-		
-		MediaType payloadContentType = MediaType.TEXT_PLAIN;
-		
-		if(responsePayload != null && messageUtil.isValidJSON(responsePayload)) {
-			payloadContentType = MediaType.APPLICATION_JSON;
-		}
-		
-		// returns John Doe if everything is OK
-		if (!("ids:RejectionMessage".equals(responseHeaders.get("IDS-Messagetype").get(0)))) {
-			response = ResponseEntity.ok()
-					.headers(responseHeaders)
-					.contentType(payloadContentType)
-					.body(responsePayload);
-		}
-		
-		return response;
-	}
+		HttpHeaders responseHeaders = new HttpHeaders();
 
-	private HttpHeaders createResponseMessageHeaders(HttpHeaders httpHeaders, String rejectionReason) {
-		String requestMessageType = httpHeaders.getFirst("IDS-Messagetype");
-		HttpHeaders headers = new HttpHeaders();
+		Message message = HttpHeadersUtil.httpHeadersToMessage(httpHeaders);
+		// Create handler based on type of message and get map with header and payload
+		DataAppMessageHandler handler = factory.createMessageHandler(message.getClass());
+		Map<String, Object> responseMap = handler.handleMessage(message, payload);
+		Map<String, Object> responseHeaderMap = new HashMap<>();
+		if (responseMap.get("header") != null) {
+			responseHeaderMap = HttpHeadersUtil.messageToHttpHeaders((Message) responseMap.get("header"));
```

Comment:
Can we create/update HttpHeaderUtil method to create HttpHeader directly, not to have to convert from Map to HttpHeaders?
We can have both methods, if we need IDS message -> Map<String, Object>

Reply:
Nice tip, I left the old method and renamed it to messageToMap(), and created a new one that converts messages directly to httpHeaders.